File: app/lib/data/repositories/profile_repository.dart
import 'dart:convert';
import 'package:drift/drift.dart';
import '../db/database.dart';
import '../../domain/models/user_profile.dart';

class ProfileRepository {
  final AppDatabase db;
  ProfileRepository(this.db);

  Future<UserProfile?> get() async {
    final row = await (db.select(db.userProfiles)..limit(1)).getSingleOrNull();
    if (row == null) return null;
    return UserProfile(
      age: row.age,
      heightCm: row.heightCm,
      weightKg: row.weightKg,
      sex: Sex.values.byName(row.sex),
      goal: Goal.values.byName(row.goal),
      experience: Experience.values.byName(row.experience),
      equipment: (jsonDecode(row.equipment) as List).cast<String>(),
      injuries: row.injuries,
      dietaryNotes: row.dietaryNotes,
      dailySchedule:
          (jsonDecode(row.dailySchedule) as Map).cast<String, bool>(),
      updatedAt: row.updatedAt,
    );
  }

  Future<void> save(UserProfile p) async {
    await db.into(db.userProfiles).insert(
      UserProfilesCompanion.insert(
        age: p.age,
        heightCm: p.heightCm,
        weightKg: p.weightKg,
        sex: p.sex.name,
        goal: p.goal.name,
        experience: p.experience.name,
        equipment: jsonEncode(p.equipment),
        injuries: Value(p.injuries),
        dietaryNotes: Value(p.dietaryNotes),
        dailySchedule: jsonEncode(p.dailySchedule),
        updatedAt: p.updatedAt,
      ),
    );
  }
}

File: app/android/app/src/main/java/com/gymcoach/gym_coach/data/repository/ProfileRepository.kt
package com.gymcoach.gym_coach.data.repository

import com.gymcoach.gym_coach.data.db.dao.UserProfileDao
import com.gymcoach.gym_coach.data.db.entity.UserProfileEntity
import com.gymcoach.gym_coach.domain.model.Experience
import com.gymcoach.gym_coach.domain.model.Goal
import com.gymcoach.gym_coach.domain.model.Sex
import com.gymcoach.gym_coach.domain.model.UserProfile
import kotlinx.serialization.builtins.ListSerializer
import kotlinx.serialization.builtins.MapSerializer
import kotlinx.serialization.builtins.serializer
import kotlinx.serialization.json.Json
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Mirrors `lib/data/repositories/profile_repository.dart`. Note `save` is an
 * INSERT, never an upsert; calling it twice yields two rows and `get()` will
 * only return the first.
 */
@Singleton
class ProfileRepository @Inject constructor(
    private val dao: UserProfileDao
) {
    private val json = Json { ignoreUnknownKeys = true }

    suspend fun get(): UserProfile? {
        val row = dao.first() ?: return null
        return UserProfile(
            age = row.age,
            heightCm = row.heightCm,
            weightKg = row.weightKg,
            sex = Sex.valueOf(row.sex),
            goal = Goal.valueOf(row.goal),
            experience = Experience.valueOf(row.experience),
            equipment = json.decodeFromString(ListSerializer(String.serializer()), row.equipment),
            injuries = row.injuries,
            dietaryNotes = row.dietaryNotes,
            dailySchedule = if (row.dailySchedule.isEmpty()) emptyMap()
            else json.decodeFromString(MapSerializer(String.serializer(), Boolean.serializer()), row.dailySchedule),
            updatedAt = row.updatedAt
        )
    }

    suspend fun save(profile: UserProfile): Long {
        val entity = UserProfileEntity(
            age = profile.age,
            heightCm = profile.heightCm,
            weightKg = profile.weightKg,
            sex = profile.sex.name,
            goal = profile.goal.name,
            experience = profile.experience.name,
            equipment = json.encodeToString(ListSerializer(String.serializer()), profile.equipment),
            injuries = profile.injuries,
            dietaryNotes = profile.dietaryNotes,
            dailySchedule = json.encodeToString(
                MapSerializer(String.serializer(), Boolean.serializer()),
                profile.dailySchedule
            ),
            updatedAt = profile.updatedAt
        )
        return dao.insert(entity)
    }
}
